From 2799775899d630d211647d5c916b987746dd7fbf Mon Sep 17 00:00:00 2001
From: Alexander Sosedkin <asosedkin@redhat.com>
Date: Fri, 10 Apr 2026 20:30:27 +0200
Subject: [PATCH] lib/x509: fix cleanup when gnutls_x509_crt_list_import_pkcs11
 fails

Previously gnutls_x509_trust_list_remove_trust_file tried to free
the entire xcrt_list, even though one source of failures is
gnutls_pkcs11_obj_list_import_url2 that deinits it up to the correct
position.

With this change, both functions zero unused entries.

Reported-by: Joshua Rogers of AISLE Research Team <joshua@joshua.hu>
Fixes: #1819
Signed-off-by: Alexander Sosedkin <asosedkin@redhat.com>
---
 lib/pkcs11.c            | 1 +
 lib/x509/verify-high2.c | 4 ++--
 2 files changed, 3 insertions(+), 2 deletions(-)

--- a/lib/pkcs11.c
+++ b/lib/pkcs11.c
@@ -3732,10 +3732,11 @@ gnutls_x509_crt_list_import_pkcs11(gnutl
 	return 0;
 
       cleanup:
 	for (j = 0; j < i; j++) {
 		gnutls_x509_crt_deinit(certs[j]);
+		certs[j] = NULL;
 	}
 
 	return ret;
 }
 
--- a/lib/x509/verify-high2.c
+++ b/lib/x509/verify-high2.c
@@ -214,12 +214,11 @@ int add_trust_list_pkcs11_object_url(gnu
 	if (pcrt_list_size == 0) {
 		ret = 0;
 		goto cleanup;
 	}
 
-	xcrt_list = _gnutls_reallocarray(NULL, pcrt_list_size,
-					 sizeof(gnutls_x509_crt_t));
+	xcrt_list = gnutls_calloc(pcrt_list_size, sizeof(gnutls_x509_crt_t));
 	if (xcrt_list == NULL) {
 		ret = GNUTLS_E_MEMORY_ERROR;
 		goto cleanup;
 	}
 
